aws-api-appsync/src/main/java/com/amplifyframework/datastore/appsync/SerializedModelAdapter.java

```diff
@@ -0,0 +1,98 @@
+import com.google.gson.JsonDeserializationContext;
+import com.google.gson.JsonDeserializer;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParseException;
+import com.google.gson.JsonPrimitive;
+import com.google.gson.JsonSerializationContext;
+import com.google.gson.JsonSerializer;
+
+import java.lang.reflect.Type;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Deserializer for SerializedModel. Helpful to deserialize from the graphql response.
+ */
+public final class SerializedModelAdapter
+        implements JsonDeserializer<SerializedModel>, JsonSerializer<SerializedModel> {
+    private SerializedModelAdapter() {}
+
+    /**
+     * Registers an adapter with a Gson builder.
+     * @param builder A gson builder
+     */
+    public static void register(GsonBuilder builder) {
+        builder.registerTypeAdapter(SerializedModel.class, new SerializedModelAdapter());
+    }
+
+    @Override
+    public JsonElement serialize(SerializedModel src, Type typeOfSrc, JsonSerializationContext context) {
+        ModelSchema schema = src.getModelSchema();
+
+        JsonObject result = new JsonObject();
+        result.add("id", context.serialize(src.getId()));
+        result.add("modelSchema", context.serialize(schema));
+
+        JsonObject serializedData = new JsonObject();
+        for (Map.Entry<String, Object> entry : src.getSerializedData().entrySet()) {
+            if (entry.getValue() instanceof SerializedModel) {
+                SerializedModel hpm = (SerializedModel) entry.getValue();
+                serializedData.add(entry.getKey(), new JsonPrimitive(hpm.getId()));
+            } else {
+                serializedData.add(entry.getKey(), context.serialize(entry.getValue()));
+            }
+        }
+        result.add("serializedData", serializedData);
+        return result;
+    }
+
+    @Override
+    public SerializedModel deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context)
+            throws JsonParseException {
+        JsonObject object = json.getAsJsonObject();
+        ModelSchema modelSchema = context.deserialize(object.get("modelSchema"), ModelSchema.class);
+
+        JsonObject serializedDataObject = object.get("serializedData").getAsJsonObject();
+        Map<String, Object> serializedData = new HashMap<>(GsonObjectConverter.toMap(serializedDataObject));
+
+        // Patch up nested models as HybridPlatformModel themselves.
+        for (Map.Entry<String, JsonElement> item : serializedDataObject.entrySet()) {
+            ModelField field = modelSchema.getFields().get(item.getKey());
+            if (field != null && field.isModel()) {
+                serializedData.put(field.getName(), SerializedModel.builder()
+                    .serializedData(Collections.singletonMap("id", item.getValue().getAsString()))
+                    .modelSchema(null)
+                    .build());
```

Comment:
I think that would be an improvement, yes. It would open the door for an infinite loop, though, wouldn't it? Regardless, I think I'd prefer to tackle it in future work. If it's amenable to you, I'll leave a `TODO` note here in the code, to capture the future work.

Reply:
> It would open the door for an infinite loop, though, wouldn't it?

Hmm, we'd need to make sure that is avoided.  Given a `Blog`, and a `BlogOwner`, they both reference each other, but the `Blog` is the "owner" in the relationship.  We can check this via `ModelAssociation::isOwner`

Getting that all working is non-trivial, so tackling at a later date sounds good to me.